File: app/src/main/java/com/youer/genshin/ui/MainFragment.java
package com.youer.genshin.ui;

import java.util.ArrayList;
import java.util.List;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.youer.genshin.App;
import com.youer.genshin.databinding.FragmentMainBinding;
import com.youer.genshin.req.CalculateReq;
import com.youer.genshin.req.CalculateReq.CharacterInfo;
import com.youer.genshin.resp.CalculateResp;
import com.youer.genshin.resp.Result;
import com.youer.genshin.view.CharacterDialog;
import com.youer.genshin.view.CharacterDialog.ConfirmListener;
import com.youer.genshin.view.CharacterInfoView;
import com.youer.genshin.view.CharacterInfoView.DeleteListener;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * @author youer
 * @date 9/20/23
 */
public class MainFragment extends BaseFragment {

    private static final String TAG = "MainFragment";
    CalculateReq calculateReq;
    FragmentMainBinding binding;
    private ExpandAdapter adapter;
    List<CharacterInfo> characterInfoList;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentMainBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        characterInfoList = new ArrayList<>();
        calculateReq = new CalculateReq();
        calculateReq.setUid(uid);
        calculateReq.setCharacters(characterInfoList);
        adapter = new ExpandAdapter();
        binding.list.setAdapter(adapter);
        binding.add.setOnClickListener(addListener);
        binding.calculate.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                calculate();
            }
        });
    }

    public OnClickListener addListener = new OnClickListener() {

        @Override
        public void onClick(View v) {
            CharacterDialog dialog = new CharacterDialog(v.getContext());
            dialog.setConfirmListener(new ConfirmListener() {
                @Override
                public void onConfirm(CharacterInfo characterInfo) {
                    characterInfoList.add(characterInfo);
                    CharacterInfoView view = new CharacterInfoView(getContext(), characterInfo);
                    view.setListener(new DeleteListener() {
                        @Override
                        public void onDelete(View view, CharacterInfo characterInfo) {
                            characterInfoList.remove(characterInfo);
                            binding.llCharacter.removeView(view);
                        }
                    });
                    binding.llCharacter.addView(view);
                }
            });
            dialog.show();
        }
    };

    public void calculate() {
        if (characterInfoList.isEmpty()) {
            Toast.makeText(getContext(), "请先添加角色", Toast.LENGTH_SHORT).show();
            return;
        }
        Call<Result<CalculateResp>> call = App.getApiService().calculateCharacterRelics(calculateReq);
        call.enqueue(new Callback<Result<CalculateResp>>() {
            @Override
            public void onResponse(Call<Result<CalculateResp>> call, Response<Result<CalculateResp>> response) {
                if (response.isSuccessful()) {
                    Result<CalculateResp> result = response.body();
                    if (result.isSuccess()) {
                        CalculateResp resp = result.getResult();
                        adapter.setData(resp.getCharacters());
                    }
                }
            }

            @Override
            public void onFailure(Call<Result<CalculateResp>> call, Throwable t) {
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }

}

File: app/src/main/java/com/youer/genshin/ui/MainFragment.kt
package com.youer.genshin.ui

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.youer.genshin.App
import com.youer.genshin.databinding.FragmentMainBinding
import com.youer.genshin.presenter.MainPresenter
import com.youer.genshin.presenter.ResultCallback
import com.youer.genshin.req.CalculateReq
import com.youer.genshin.req.CalculateReq.CharacterInfo
import com.youer.genshin.resp.CalculateResp
import com.youer.genshin.resp.Result
import com.youer.genshin.view.CharacterDialog
import com.youer.genshin.view.CharacterDialog.ConfirmListener
import com.youer.genshin.view.CharacterInfoView
import com.youer.genshin.view.CharacterInfoView.DeleteListener
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import kotlin.collections.ArrayList

/**
 * @author youer
 * @date 9/20/23
 */
class MainFragment(var presenter: MainPresenter) : BaseFragment(presenter) {
    private var calculateReq = CalculateReq()
    private lateinit var binding: FragmentMainBinding
    private var adapter: ExpandAdapter = ExpandAdapter()
    private var characterInfoList = ArrayList<CharacterInfo>()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentMainBinding.inflate(inflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        calculateReq.uid= uid
        calculateReq.characters = characterInfoList
        binding.list.setAdapter(adapter)
        binding.add.setOnClickListener(addListener)
        binding.calculate.setOnClickListener { calculate() }
    }

    private var addListener = View.OnClickListener { v ->
        CharacterDialog(v.context).apply {
            confirmListener = object : ConfirmListener {
                override fun onConfirm(characterInfo: CharacterInfo) {
                    characterInfoList.add(characterInfo)
                    CharacterInfoView(context, characterInfo)
                            .apply {
                                listener = object : DeleteListener {
                                    override fun onDelete(view: View, characterInfo: CharacterInfo) {
                                        characterInfoList.remove(characterInfo)
                                        binding.llCharacter.removeView(view)
                                    }
                                }
                                binding.llCharacter.addView(this)
                            }

                }

            }
            show()
        }
    }

    fun calculate() {
        if (calculateReq.characters.isEmpty()) {
            Toast.makeText(context, "请先添加角色", Toast.LENGTH_SHORT).show()
            return
        }
        presenter.calculate(calculateReq,object : ResultCallback<List<CalculateResp.CharacterResp>?>{
            override fun success(t: List<CalculateResp.CharacterResp>?) {
                adapter.setData(t)
            }

            override fun fail(s: String) {
                Toast.makeText(context, s, Toast.LENGTH_SHORT).show()
            }
        })
    }

    companion object {
        private const val TAG = "MainFragment"
    }
}